Userland/Libraries/LibSQL/Value.cpp

```diff
@@ -1,301 +1,340 @@
+SQLType Value::type() const
 {
-    return Value(Value::SetImplementationSingleton, ArrayImpl(element_type, max_size));
+    return m_type;
 }
 
-Value const& Value::null()
+StringView Value::type_name() const
 {
-    static Value s_null(SQLType::Null);
-    return s_null;
+    switch (type()) {
+#undef __ENUMERATE_SQL_TYPE
+#define __ENUMERATE_SQL_TYPE(name, cardinal, type, impl, size) \
+    case SQLType::type:                                        \
+        return name##sv;
+        ENUMERATE_SQL_TYPES(__ENUMERATE_SQL_TYPE)
+#undef __ENUMERATE_SQL_TYPE
+    default:
+        VERIFY_NOT_REACHED();
+    }
 }
 
 bool Value::is_null() const
 {
-    return m_impl.visit([&](auto& impl) { return impl.is_null(); });
+    return !m_value.has_value();
 }
 
-SQLType Value::type() const
+String Value::to_string() const
 {
-    return m_impl.visit([&](auto& impl) { return impl.type(); });
-}
+    if (is_null())
+        return "(null)"sv;
 
-String Value::type_name() const
-{
-    return m_impl.visit([&](auto& impl) { return impl.type_name(); });
-}
+    return m_value->visit(
+        [](String const& value) -> String { return value; },
+        [](int value) -> String { return String::number(value); },
+        [](double value) -> String { return String::number(value); },
+        [](bool value) -> String { return value ? "true"sv : "false"sv; },
+        [](TupleValue const& value) -> String {
+            StringBuilder builder;
 
-BaseTypeImpl Value::downcast_to_basetype() const
-{
-    return m_impl.downcast<NullImpl, TextImpl, IntegerImpl, FloatImpl, BooleanImpl>();
-}
+            builder.append('(');
+            builder.join(',', value.values);
+            builder.append(')');
 
-String Value::to_string() const
-{
-    if (is_null())
-        return "(null)";
-    return m_impl.visit([&](auto& impl) { return impl.to_string(); });
+            return builder.build();
+        });
 }
 
 Optional<int> Value::to_int() const
 {
     if (is_null())
         return {};
-    return m_impl.visit([&](auto& impl) { return impl.to_int(); });
+
+    return m_value->visit(
+        [](String const& value) -> Optional<int> { return value.to_int(); },
+        [](int value) -> Optional<int> { return value; },
+        [](double value) -> Optional<int> {
+            if (value > static_cast<double>(NumericLimits<int>::max()))
+                return {};
+            if (value < static_cast<double>(NumericLimits<int>::min()))
+                return {};
+            return static_cast<int>(round(value));
+        },
+        [](bool value) -> Optional<int> { return static_cast<int>(value); },
+        [](TupleValue const&) -> Optional<int> { return {}; });
 }
 
 Optional<u32> Value::to_u32() const
 {
-    if (is_null())
-        return {};
-    auto ret = to_int();
-    if (ret.has_value())
-        return static_cast<u32>(ret.value());
+    if (auto result = to_int(); result.has_value())
+        return static_cast<u32>(result.value());
```

Comment:
How about negative values?

Reply:
I guess it's an existing bug, so maybe just leave a note?

Reply:
Yeah, there's a few suspect signed/unsigned things in this file I marked with FIXMEs. Slapped another one here.